### DISCUSSION TITLE
Allow users to provide their own equality operator implementation in records

### DISCUSSION THREAD

**Discussion by hez2010** [dotnet/csharplang#5888] [Language Ideas]
I'm implementing an expression calculator, and I want to use `==` and `!=` to build equ exprs and neq exprs.

```csharp
public static Expr<T> operator ==(Expr<T> left, Expr<T> right) => new BinaryExpr(new BinaryOperator(Operators.Eq), left, right);
public static Expr<T> operator !=(Expr<T> left, Expr<T> right) => new BinaryExpr(new BinaryOperator(Operators.Ne), left, right);
```

So I can use `Expr<T>` like:

```csharp
Expr<int> left = ...;
Expr<int> right = ...;
Expr<int> equal = left == right;
Console.WriteLine(equal.Eval());
```

At the first time I used records for my `Expr<T>` implementation since it's convenient, and with records I don't need to provide my own `ToString` implementation. But soon I found that I cannot overload `==` as well as `!=` so I had to change those records to classes.

For the full implementation, see https://github.com/hez2010/PatternMatchingExpr/blob/main/Expr.cs.

`==` doesn't have to be the same semantic with `Equals`, and it doesn't have to return a boolean value either. Sometimes we want to overload `==` to achieve shorthand, eg. for expression building. In this case, `==` is used for building equality expression, while `Equals` is used for determining whether two expressions are the same (the generated `Equals` from records actually implements it automatically for me, but since I'm not able to overload `==` in records, I can't use records and have to use classes instead). 

I hope C# can allow users to provide the own implementation of `==` and `!=` in records.

**Comment by HaloFour** [comment]
Why are you trying to use records here?  Their purpose is to provide value equality, and it sounds like that's not what you want.

**Reply by hez2010** [reply]
Records can implement `Equals` for value equality automatically, and `ToString` for print expressions, as well as `Deconstruct` for deconstruction, and it can also save boilerplate code such as property getter and setter and constructors. I think records are pretty suitable for this scenario, except `==` and `!=` cannot be overloaded.

**Reply by HaloFour** [reply]
They do that for convenience, but the _purpose_ of records is to provide value equality and they intentionally take an opinionated stance here.  This is also why you cannot override `bool Equals(object)`.

**Reply by hez2010** [reply]
I just read about the decision meeting around equality operators in records: https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-07-27.md#equality-operators-in-records

> Finally, we looked at whether the user will be able to provide their own implementation of == and != operators, and if we'd error on this or allow it and not generate the operators ourselves in this case. We feel that allowing this would complicate records: there are existing extension points into record equality that can be overridden as needed, and we want to have a stronger concept of Equals and == being the same. If there are good user scenarios around allowing this, we can consider it at a later point.

It turns out there wasn't a user scenario at that time. Now we get a scenario for this so I think this limitation should be relaxed.

**Reply by HaloFour** [reply]
That's fair, I guess the LDM can weigh how they want records to extend outside of value equality scenarios if this is one that they want to support.
